File: nim_eavt/resolver.nim
## resolver.nim — Nim Resolver (schema cache, ID allocation).
##
## Port of spier-transactor/src/resolver.rs (~340 lines Rust → Nim).
## Manages attribute registry, partition counters, and schema metadata.

import std/[tables, sets, strutils, options]

# ═══════════════════════════════════════════════════════════════════════════════
# Constants (from resolver_consts.rs)
# ═══════════════════════════════════════════════════════════════════════════════

const
  BootstrapFirstUserId* = 100'u64

  DbIdentAid* = 1'u32
  DbCardinalityAid* = 2'u32
  DbValueTypeAid* = 3'u32
  DbUniqueAid* = 5'u32
  DbIndexAid* = 6'u32
  DbTxInstantAid* = 9'u32
  DbPartIdAid* = 39'u32

  DbTypeString* = 20'u32
  DbTypeRef* = 21'u32
  DbTypeLong* = 22'u32
  DbTypeKeyword* = 23'u32
  DbTypeBoolean* = 24'u32
  DbTypeInstant* = 25'u32
  DbTypeBytes* = 26'u32
  DbTypeFloat* = 27'u32
  DbTypeBlob* = 28'u32

  DbCardinalityOneAid* = 35'u32
  DbCardinalityManyAid* = 36'u32
  DbUniqueValueAid* = 37'u32
  DbUniqueIdentityAid* = 38'u32

  PartDb* = 0'u64
  PartTx* = 3'u64
  PartUser* = 4'u64

  FirstCustomPartition = 64'u64
  PartitionShift = 44'u32
  SeqMask = 0xFFFFFFFFFFF'u64

# ═══════════════════════════════════════════════════════════════════════════════
# Entity ID helpers
# ═══════════════════════════════════════════════════════════════════════════════

proc partitionOf*(eid: int64): uint64 = cast[uint64](eid) shr PartitionShift
proc seqOf*(eid: int64): int64 = cast[int64](cast[uint64](eid) and SeqMask)
proc makeEntityId*(partitionId: uint64, seq: int64): int64 =
  cast[int64]((partitionId shl PartitionShift) or cast[uint64](seq))

# ═══════════════════════════════════════════════════════════════════════════════
# Attribute name normalization
# ═══════════════════════════════════════════════════════════════════════════════

proc normalizeAttr*(name: string): string =
  if name.len > 0 and name[0] == ':' and '/' in name:
    result = name[1..^1].replace('/', '.')
  elif '.' notin name:
    raise newException(ValueError,
      "attribute name must include namespace (e.g. 'company.name'), got " & name)
  else:
    result = name

# ═══════════════════════════════════════════════════════════════════════════════
# Bootstrap schema
# ═══════════════════════════════════════════════════════════════════════════════

const BootstrapSchema: array[26, (string, uint32)] = [
  ("db.ident", 1),
  ("db.cardinality", 2),
  ("db.valueType", 3),
  ("db.isComponent", 4),
  ("db.unique", 5),
  ("db.index", 6),
  ("db.fulltext", 7),
  ("db.noHistory", 8),
  ("db.txInstant", 9),
  ("db.type.string", 20),
  ("db.type.ref", 21),
  ("db.type.long", 22),
  ("db.type.keyword", 23),
  ("db.type.boolean", 24),
  ("db.type.instant", 25),
  ("db.type.bytes", 26),
  ("db.type.float", 27),
  ("db.type.blob", 28),
  ("db.cardinality.one", 35),
  ("db.cardinality.many", 36),
  ("db.unique.value", 37),
  ("db.unique.identity", 38),
  ("db.part/id", 39),
  ("db.part/db", 40),
  ("db.part/tx", 41),
  ("db.part/user", 42),
]

# ═══════════════════════════════════════════════════════════════════════════════
# Resolver
# ═══════════════════════════════════════════════════════════════════════════════

type
  PartitionCounter = object
    nextSeq: int64

  Resolver* = object
    attrs: Table[string, uint32]
    attrsRev: Table[uint32, string]
    nextAid: uint32
    partitions: Table[uint64, PartitionCounter]
    partitionNames: Table[string, uint64]
    nextCustomPartition: uint64
    cardinality: Table[uint32, bool]
    declared: HashSet[uint32]
    valueTypes: Table[uint32, uint32]
    uniqueAttrs: HashSet[uint32]
    indexedAttrs: HashSet[uint32]
    ## Declaration order (attr names by ascending aid; partition names by
    ## ascending id). Needed to replay the schema into the Rust compiler
    ## engine so it assigns the SAME ids.
    attrDeclOrder*: seq[string]
    partDeclOrder*: seq[string]

proc newResolver*(): Resolver =
  result.nextAid = 1
  result.nextCustomPartition = FirstCustomPartition
  result.attrs = initTable[string, uint32]()
  result.attrsRev = initTable[uint32, string]()
  result.partitions = initTable[uint64, PartitionCounter]()
  result.partitionNames = initTable[string, uint64]()
  result.cardinality = initTable[uint32, bool]()
  result.declared = initHashSet[uint32]()
  result.valueTypes = initTable[uint32, uint32]()
  result.uniqueAttrs = initHashSet[uint32]()
  result.indexedAttrs = initHashSet[uint32]()

  for (name, aid) in BootstrapSchema:
    result.attrs[name] = aid
    result.attrsRev[aid] = name
    result.declared.incl aid
    if aid >= result.nextAid:
      result.nextAid = aid + 1

  result.valueTypes[DbValueTypeAid] = DbTypeRef
  result.valueTypes[DbCardinalityAid] = DbTypeRef
  result.valueTypes[DbUniqueAid] = DbTypeRef
  result.valueTypes[DbIdentAid] = DbTypeString
  result.valueTypes[DbPartIdAid] = DbTypeLong
  result.valueTypes[DbTxInstantAid] = DbTypeInstant
  result.indexedAttrs.incl DbIdentAid

  result.partitions[PartDb] = PartitionCounter(nextSeq: BootstrapFirstUserId.int64)
  result.partitionNames["db.part/db"] = PartDb
  result.partitions[PartTx] = PartitionCounter(nextSeq: 1)
  result.partitionNames["db.part/tx"] = PartTx
  result.partitions[PartUser] = PartitionCounter(nextSeq: 1)
  result.partitionNames["db.part/user"] = PartUser

# ── Partition management (must come before attr resolution — uses forward ref) ──

proc allocateInPartition*(r: var Resolver; partitionId: uint64): int64 =
  if partitionId notin r.partitions:
    raise newException(ValueError, "unknown partition: " & $partitionId)
  let seqVal = r.partitions[partitionId].nextSeq
  r.partitions[partitionId].nextSeq = seqVal + 1
  return makeEntityId(partitionId, cast[int64](seqVal))

proc allocateEntityId*(r: var Resolver): int64 =
  allocateInPartition(r, PartUser)

proc allocateSchemaId*(r: var Resolver): int64 =
  allocateInPartition(r, PartDb)

proc partitionIdFor*(r: var Resolver; name: string): Option[uint64] =
  if name in r.partitionNames: return some(r.partitionNames[name])

proc declarePartition*(r: var Resolver; name: string): uint64 =
  if name in r.partitionNames: return r.partitionNames[name]
  let p = r.nextCustomPartition
  inc r.nextCustomPartition
  r.partitions[p] = PartitionCounter(nextSeq: 1)
  r.partitionNames[name] = p
  r.partDeclOrder.add name
  return p

proc registerPartition*(r: var Resolver; name: string; partitionId: uint64) =
  if name in r.partitionNames: return
  if partitionId notin r.partitions:
    r.partitions[partitionId] = PartitionCounter(nextSeq: 1)
  r.partitionNames[name] = partitionId
  if partitionId >= FirstCustomPartition and partitionId >= r.nextCustomPartition:
    r.nextCustomPartition = partitionId + 1

proc defaultUserPartition*(r: var Resolver): uint64 = PartUser
proc knownPartitions*(r: var Resolver): seq[uint64] =
  for k in r.partitions.keys: result.add k

# ── Attribute resolution ──

proc lookupAttr*(r: var Resolver; name: string): Option[uint32] =
  try:
    let n = normalizeAttr(name)
    if n in r.attrs: return some(r.attrs[n])
  except: discard

proc isDeclared*(r: var Resolver; aid: uint32): bool = aid in r.declared

proc internAttr*(r: var Resolver; name: string): uint32 =
  let n = normalizeAttr(name)
  if n in r.attrs: return r.attrs[n]
  let eid = allocateInPartition(r, PartDb)
  let aid = eid.uint32
  r.nextAid = aid + 1
  r.attrs[n] = aid
  r.attrsRev[aid] = n
  return aid

proc declareAttr*(r: var Resolver; name: string; valueType: uint32;
                   many: bool): (uint32, bool) =
  let n = normalizeAttr(name)
  if n in r.attrs and r.attrs[n] in r.declared:
    return (r.attrs[n], false)
  let seqVal = allocateInPartition(r, PartDb)
  let aid = seqVal.uint32
  r.nextAid = aid + 1
  r.attrs[n] = aid
  r.attrsRev[aid] = n
  r.declared.incl aid
  r.valueTypes[aid] = valueType
  if many: r.cardinality[aid] = true
  r.attrDeclOrder.add n
  return (aid, true)

proc valueTypeFor*(r: var Resolver; aid: uint32): Option[uint32] =
  if aid in r.valueTypes: return some(r.valueTypes[aid])

proc attrName*(r: var Resolver; aid: uint32): string =
  r.attrsRev.getOrDefault(aid, $aid)

proc attrNameOpt*(r: var Resolver; aid: uint32): Option[string] =
  if aid in r.attrsRev: return some(r.attrsRev[aid])

# ── Cardinality / uniqueness / indexing ──

proc isMany*(r: var Resolver; aid: uint32): bool =
  aid in r.cardinality

proc setCardinality*(r: var Resolver; aid: uint32; many: bool) =
  if many: r.cardinality[aid] = true else: r.cardinality.del(aid)

proc isUnique*(r: var Resolver; aid: uint32): bool =
  aid in r.uniqueAttrs

proc setUnique*(r: var Resolver; aid: uint32; unique: bool) =
  if unique: r.uniqueAttrs.incl aid else: r.uniqueAttrs.excl aid

proc isIndexed*(r: var Resolver; aid: uint32): bool =
  aid in r.uniqueAttrs or aid in r.indexedAttrs

proc setIndexed*(r: var Resolver; aid: uint32; indexed: bool) =
  if indexed: r.indexedAttrs.incl aid else: r.indexedAttrs.excl aid

# ── Sequence management ──

proc advancePast*(r: var Resolver; eid: int64) =
  let p = partitionOf(eid)
  let s = seqOf(eid)
  if p in r.partitions:
    if s >= r.partitions[p].nextSeq:
      r.partitions[p].nextSeq = s + 1

proc setPartitionSeq*(r: var Resolver; partitionId: uint64, seq: int64) =
  if partitionId in r.partitions:
    if seq > r.partitions[partitionId].nextSeq:
      r.partitions[partitionId].nextSeq = seq

proc nextEntId*(r: var Resolver): int64 =
  if PartDb in r.partitions: return r.partitions[PartDb].nextSeq
  return BootstrapFirstUserId.int64

# ── Batch loading ──

proc loadAttrs*(r: var Resolver; items: seq[(seq[byte], seq[byte])]) =
  for (k, v) in items:
    let name = newString(k.len)
    if k.len > 0: copyMem(addr name[0], addr k[0], k.len)
    if v.len < 4: continue
    let aid = (uint32(v[0]) shl 24 or uint32(v[1]) shl 16 or
                uint32(v[2]) shl 8 or uint32(v[3]))
    r.attrs[name] = aid
    r.attrsRev[aid] = name
    if aid >= r.nextAid: r.nextAid = aid + 1

proc loadUserAttr*(r: var Resolver; name: string; eid: int64;
                    valueType: uint32; many, unique, indexed: bool) =
  let aid = eid.uint32
  let isNew = aid notin r.declared
  r.attrs[name] = aid
  r.attrsRev[aid] = name
  r.declared.incl aid
  r.valueTypes[aid] = valueType
  if many: r.cardinality[aid] = true
  if unique: r.uniqueAttrs.incl aid
  if indexed: r.indexedAttrs.incl aid
  if isNew:
    # keep attrDeclOrder sorted by aid (= declaration order in PartDb)
    var insPos = r.attrDeclOrder.len
    for i, n in r.attrDeclOrder:
      if r.attrs.getOrDefault(n, 0'u32) > aid: insPos = i; break
    r.attrDeclOrder.insert(name, insPos)
  let p = partitionOf(eid)
  if p in r.partitions:
    if seqOf(eid) >= r.partitions[p].nextSeq:
      r.partitions[p].nextSeq = seqOf(eid) + 1
  if aid >= r.nextAid: r.nextAid = aid + 1

File: py_eavt/src/eavt/resolver.py
"""resolver.py — Schema cache, ID allocation.

Port of nim_eavt/resolver.nim. Manages attribute registry, partition counters,
and schema metadata.
"""
from __future__ import annotations

from .types import (
    BOOTSTRAP_SCHEMA,
    BOOTSTRAP_FIRST_USER_ID,
    DB_CARDINALITY_AID,
    DB_CARDINALITY_MANY,
    DB_CARDINALITY_ONE,
    DB_IDENT_AID,
    DB_INDEX_AID,
    DB_IS_COMPONENT_AID,
    DB_NO_HISTORY_AID,
    DB_TX_INSTANT_AID,
    DB_TYPE_BOOLEAN,
    DB_TYPE_INSTANT,
    DB_TYPE_LONG,
    DB_TYPE_REF,
    DB_TYPE_STRING,
    DB_UNIQUE_AID,
    DB_UNIQUE_IDENTITY,
    DB_UNIQUE_VALUE,
    DB_VALUE_TYPE_AID,
    DB_PART_ID_AID,
    FIRST_CUSTOM_PARTITION,
    PARTITION_SHIFT,
    PART_DB,
    PART_TX,
    PART_USER,
    SEQ_MASK,
)


# ═══════════════════════════════════════════════════════════════════════════════
# Entity ID helpers
# ═══════════════════════════════════════════════════════════════════════════════


def partition_of(eid: int) -> int:
    return (eid & 0xFFFFFFFFFFFFFFFF) >> PARTITION_SHIFT


def seq_of(eid: int) -> int:
    return eid & SEQ_MASK


def make_entity_id(partition_id: int, seq: int) -> int:
    return ((partition_id & 0xFFFFFFFFFF) << PARTITION_SHIFT) | (seq & SEQ_MASK)


# ═══════════════════════════════════════════════════════════════════════════════
# Attribute name normalization
# ═══════════════════════════════════════════════════════════════════════════════


def normalize_attr(name: str) -> str:
    if name.startswith(":") and "/" in name:
        return name[1:].replace("/", ".")
    if "." not in name:
        raise ValueError(f"attribute name must include namespace (e.g. 'company.name'), got {name}")
    return name


# ═══════════════════════════════════════════════════════════════════════════════
# Resolver
# ═══════════════════════════════════════════════════════════════════════════════


class PartitionCounter:
    __slots__ = ("next_seq",)

    def __init__(self, next_seq: int = 1):
        self.next_seq = next_seq


class Resolver:
    __slots__ = (
        "attrs",
        "attrs_rev",
        "next_aid",
        "partitions",
        "partition_names",
        "next_custom_partition",
        "cardinality",
        "declared",
        "value_types",
        "unique_attrs",
        "indexed_attrs",
        "attr_decl_order",
        "part_decl_order",
    )

    def __init__(self):
        self.attrs: dict[str, int] = {}
        self.attrs_rev: dict[int, str] = {}
        self.next_aid: int = 1
        self.partitions: dict[int, PartitionCounter] = {}
        self.partition_names: dict[str, int] = {}
        self.next_custom_partition: int = FIRST_CUSTOM_PARTITION
        self.cardinality: dict[int, bool] = {}  # aid → is_many
        self.declared: set[int] = set()
        self.value_types: dict[int, int] = {}
        self.unique_attrs: set[int] = set()
        self.indexed_attrs: set[int] = set()
        self.attr_decl_order: list[str] = []
        self.part_decl_order: list[str] = []

        # Bootstrap schema
        for name, aid in BOOTSTRAP_SCHEMA:
            self.attrs[name] = aid
            self.attrs_rev[aid] = name
            self.declared.add(aid)
            if aid >= self.next_aid:
                self.next_aid = aid + 1

        # Default value types for bootstrap attrs
        self.value_types[DB_VALUE_TYPE_AID] = DB_TYPE_REF
        self.value_types[DB_CARDINALITY_AID] = DB_TYPE_REF
        self.value_types[DB_UNIQUE_AID] = DB_TYPE_REF
        self.value_types[DB_IDENT_AID] = DB_TYPE_STRING
        self.value_types[DB_PART_ID_AID] = DB_TYPE_LONG
        self.value_types[DB_TX_INSTANT_AID] = DB_TYPE_INSTANT
        self.indexed_attrs.add(DB_IDENT_AID)

        for name, aid in BOOTSTRAP_SCHEMA:
            if aid not in self.value_types:
                if name in ("db.ident", "db.part/id"):
                    vt = DB_TYPE_STRING
                elif name == "db.txInstant":
                    vt = DB_TYPE_INSTANT
                elif name in ("db.isComponent", "db.index", "db.fulltext", "db.noHistory"):
                    vt = DB_TYPE_BOOLEAN
                elif name.startswith("db.type."):
                    vt = DB_TYPE_LONG
                else:
                    vt = DB_TYPE_REF
                self.value_types[aid] = vt

        # Partitions
        self.partitions[PART_DB] = PartitionCounter(BOOTSTRAP_FIRST_USER_ID)
        self.partition_names["db.part/db"] = PART_DB
        self.partitions[PART_TX] = PartitionCounter(1)
        self.partition_names["db.part/tx"] = PART_TX
        self.partitions[PART_USER] = PartitionCounter(1)
        self.partition_names["db.part/user"] = PART_USER

    # ── Partition management ──

    def allocate_in_partition(self, partition_id: int) -> int:
        if partition_id not in self.partitions:
            raise ValueError(f"unknown partition: {partition_id}")
        seq_val = self.partitions[partition_id].next_seq
        self.partitions[partition_id].next_seq = seq_val + 1
        return make_entity_id(partition_id, seq_val)

    def allocate_entity_id(self) -> int:
        return self.allocate_in_partition(PART_USER)

    def allocate_schema_id(self) -> int:
        return self.allocate_in_partition(PART_DB)

    def partition_id_for(self, name: str) -> int | None:
        return self.partition_names.get(name)

    def declare_partition(self, name: str) -> int:
        if name in self.partition_names:
            return self.partition_names[name]
        p = self.next_custom_partition
        self.next_custom_partition += 1
        self.partitions[p] = PartitionCounter(1)
        self.partition_names[name] = p
        self.part_decl_order.append(name)
        return p

    def register_partition(self, name: str, partition_id: int):
        if name in self.partition_names:
            return
        if partition_id not in self.partitions:
            self.partitions[partition_id] = PartitionCounter(1)
        self.partition_names[name] = partition_id
        if partition_id >= FIRST_CUSTOM_PARTITION and partition_id >= self.next_custom_partition:
            self.next_custom_partition = partition_id + 1

    def default_user_partition(self) -> int:
        return PART_USER

    def known_partitions(self) -> list[int]:
        return list(self.partitions.keys())

    # ── Attribute resolution ──

    def lookup_attr(self, name: str) -> int | None:
        try:
            n = normalize_attr(name)
            return self.attrs.get(n)
        except ValueError:
            return None

    def is_declared(self, aid: int) -> bool:
        return aid in self.declared

    def intern_attr(self, name: str) -> int:
        n = normalize_attr(name)
        if n in self.attrs:
            return self.attrs[n]
        eid = self.allocate_in_partition(PART_DB)
        aid = eid & 0xFFFFFFFF
        self.next_aid = aid + 1
        self.attrs[n] = aid
        self.attrs_rev[aid] = n
        return aid

    def declare_attr(self, name: str, value_type: int, many: bool) -> tuple[int, bool]:
        n = normalize_attr(name)
        if n in self.attrs and self.attrs[n] in self.declared:
            return self.attrs[n], False
        seq_val = self.allocate_in_partition(PART_DB)
        aid = seq_val & 0xFFFFFFFF
        self.next_aid = aid + 1
        self.attrs[n] = aid
        self.attrs_rev[aid] = n
        self.declared.add(aid)
        self.value_types[aid] = value_type
        if many:
            self.cardinality[aid] = True
        self.attr_decl_order.append(n)
        return aid, True

    def value_type_for(self, aid: int) -> int | None:
        return self.value_types.get(aid)

    def attr_name(self, aid: int) -> str:
        return self.attrs_rev.get(aid, str(aid))

    def attr_name_opt(self, aid: int) -> str | None:
        return self.attrs_rev.get(aid)

    # ── Cardinality / uniqueness / indexing ──

    def is_many(self, aid: int) -> bool:
        return aid in self.cardinality

    def set_cardinality(self, aid: int, many: bool):
        if many:
            self.cardinality[aid] = True
        else:
            self.cardinality.pop(aid, None)

    def is_unique(self, aid: int) -> bool:
        return aid in self.unique_attrs

    def set_unique(self, aid: int, unique: bool):
        if unique:
            self.unique_attrs.add(aid)
        else:
            self.unique_attrs.discard(aid)

    def is_indexed(self, aid: int) -> bool:
        return aid in self.unique_attrs or aid in self.indexed_attrs

    def set_indexed(self, aid: int, indexed: bool):
        if indexed:
            self.indexed_attrs.add(aid)
        else:
            self.indexed_attrs.discard(aid)

    # ── Sequence management ──

    def advance_past(self, eid: int):
        p = partition_of(eid)
        s = seq_of(eid)
        if p in self.partitions:
            if s >= self.partitions[p].next_seq:
                self.partitions[p].next_seq = s + 1

    def set_partition_seq(self, partition_id: int, seq: int):
        if partition_id in self.partitions:
            if seq > self.partitions[partition_id].next_seq:
                self.partitions[partition_id].next_seq = seq

    def next_ent_id(self) -> int:
        if PART_DB in self.partitions:
            return self.partitions[PART_DB].next_seq
        return BOOTSTRAP_FIRST_USER_ID

    # ── Batch loading ──

    def load_user_attr(
        self, name: str, eid: int, value_type: int, many: bool, unique: bool, indexed: bool
    ):
        aid = eid & 0xFFFFFFFF
        is_new = aid not in self.declared
        self.attrs[name] = aid
        self.attrs_rev[aid] = name
        self.declared.add(aid)
        self.value_types[aid] = value_type
        if many:
            self.cardinality[aid] = True
        if unique:
            self.unique_attrs.add(aid)
        if indexed:
            self.indexed_attrs.add(aid)
        if is_new:
            # Insert sorted by aid
            ins_pos = len(self.attr_decl_order)
            for i, n in enumerate(self.attr_decl_order):
                if self.attrs.get(n, 0) > aid:
                    ins_pos = i
                    break
            self.attr_decl_order.insert(ins_pos, name)
        p = partition_of(eid)
        if p in self.partitions:
            if seq_of(eid) >= self.partitions[p].next_seq:
                self.partitions[p].next_seq = seq_of(eid) + 1
        if aid >= self.next_aid:
            self.next_aid = aid + 1
